### PR TITLE
Deadlocks with git dependencies

### LINKED ISSUE #2987: Deadlock when downloading git dependencies
Repro:

```
$ rm -rf ~/.cargo/registry
$ git clone git@github.com:tokio-rs/tokio.git
$ cd tokio
$ cargo metadata& 
$ cargo metadata& 

Updating registry `https://github.com/rust-lang/crates.io-index`  
Blocking waiting for file lock on the registry index 

Blocking waiting for file lock on the git checkout
Blocking waiting for file lock on the git checkout
```

### PR DESCRIPTION
**Pull request by matklad**
Only a test for now to verify that it actually fails on Travis. 

Hopefully will close #2987

### CODE REVIEW

**Comment by matklad** [comment]
So something's working!

First of all, a sanity check. When you've said

>Hm I think I like the single lock idea actually. It's pretty bad in terms of UI (makes the feature almost unusable), but it should be rarer than crates.io and it's better UI than deadlocking. So sounds good to me!

did you really meant the "single lock" approach and not "bail out if already locked" approach? Not sure if "UI" can be used to describe the former approach...

Next, to the PR itself:

Now we use a single lock for both database and checkouts. The ownership of the lock is shared across all git dependencies. Global variable is used to share the lock among git deps. It is super ugly, but I think is still a tiny bit better than using `Config` for this purpose.

**Comment by alexcrichton** [comment]
Ah yeah I did mean the single lock approach. If we bailed out with an error that'd kinda defeat the point of the file lock support in Cargo. We could pretty easily just *error* if we detect a concurrent Cargo, but the real value is figuring out what to do to sequence it all correctly.

I'd personally prefer to stick the "global lock" in a `Config` rather than a lazy static, perhaps that could be done?

**Comment by matklad** [comment]
>I'd personally prefer to stick the "global lock" in a Config rather than a lazy static, perhaps that could be done?

Done. 

>If we bailed out with an error that'd kinda defeat the point of the file lock support in Cargo.

I meant to bail out only when locking git repositories. This way you'd be able to have several truly parallel Cargo builds as long as they don't have common git dependencies. In contrast, with single lock you won't have a parallel build even if git dependencies are disjoint. But I agree that being always correct is better then being sometimes faster :)

File: src/cargo/sources/git/source.rs [hunk 1]
```diff
@@ -1,3 +1,13 @@
+//! All git sources are protected by a single file lock, which is stored
+//! in the `Config` struct. Ideally, we should use a lock per repository,
+//! but this leads to deadlocks when several instances of Cargo acquire
+//! locks in different order (see #2987).
+//!
+//! This lock must be held for the whole duration of the build and
+//! not only during checkout, because, for example, different commits
+//! from the same branch may end up in the same directory. So the lock is
+//! acquired when the first git source is updated and released when the
+//! `Config` struct is destroyed.
```
Comment (alexcrichton) [inline src/cargo/sources/git/source.rs]: Could this comment get placed on the `Config::lock_git` method instead?

Comment (matklad) [inline src/cargo/sources/git/source.rs]: Done.

**Comment by alexcrichton** [comment]
@bors: r+ 7a71d5bffca2d3b48f94e773e7b9259a8fe0ff69

Thanks!

**Pull request merged**